### PR TITLE
Push

### PR DESCRIPTION
**Pull request by ethanmick**
Added Push Notifications.
Added Social Query functions.

### CODE REVIEW

File: src/main/java/com/cloudmine/api/rest/CMURLBuilder.java [hunk 1]
```diff
@@ -38,9 +40,9 @@ public String toString() {
     }
     private static final Logger LOG = LoggerFactory.getLogger(CMURLBuilder.class);
     public static final VERSION DEFAULT_VERSION = VERSION.V1;
-    public static final String CLOUD_MINE_URL = "https://api.cloudmine.me";
+//    public static final String CLOUD_MINE_URL = "https://api.cloudmine.me";
 //    public static final String CLOUD_MINE_URL = "http://api-beta.cloudmine.me";
-//    public static final String CLOUD_MINE_URL = "http://localhost:3000";
+    public static final String CLOUD_MINE_URL = "http://10.10.20.115:3001";
```
Comment (EmilyMcCarthy) [inline src/main/java/com/cloudmine/api/rest/CMURLBuilder.java]: So I think our customers might complain about only being able to hit your box when they want to hit cloudmine

File: src/main/java/com/cloudmine/api/rest/CMURLBuilder.java [hunk 2]
```diff
@@ -201,6 +203,14 @@ private String keysToString(Collection<String> keys) {
         return keyString;
     }
 
+    public CMURLBuilder hashMapToQuery(HashMap<String, Object> map) {
```
Comment (EmilyMcCarthy) [inline src/main/java/com/cloudmine/api/rest/CMURLBuilder.java]: You don't use any hashmap specific operations, so why limit this

File: src/main/java/com/cloudmine/api/rest/CMWebService.java [hunk 3]
```diff
@@ -23,6 +24,7 @@
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
+import javax.swing.*;
```
Comment (EmilyMcCarthy) [inline src/main/java/com/cloudmine/api/rest/CMWebService.java]: I really hope we are not using Swing in here

File: src/main/java/com/cloudmine/api/rest/CMWebService.java [hunk 4]
```diff
@@ -10,6 +10,7 @@
 import com.cloudmine.api.rest.callbacks.Callback;
 import com.cloudmine.api.rest.options.CMRequestOptions;
 import com.cloudmine.api.rest.response.*;
+import com.sun.tools.internal.xjc.util.NullStream;
```
Comment (EmilyMcCarthy) [inline src/main/java/com/cloudmine/api/rest/CMWebService.java]: This seems like the sort of thing that might not be available on Android, did you check this?

File: src/main/java/com/cloudmine/api/rest/HttpVerb.java [hunk 5]
```diff
@@ -0,0 +1,12 @@
+package com.cloudmine.api.rest;
+
+/**
+ * Created with IntelliJ IDEA.
```
Comment (EmilyMcCarthy) [inline src/main/java/com/cloudmine/api/rest/HttpVerb.java]: Use the same format as the other files for this javadoc

File: src/main/java/com/cloudmine/api/rest/UserCMWebService.java [hunk 6]
```diff
@@ -111,6 +111,66 @@ public void asyncLoadAccessLists(Callback<CMObjectResponse> callback) {
...
+     * @param parameters The Parameters which will be added to the request. For example: new HashMap<String, Object>() {{ put("Param1", "Value"); }} Can be null.
+     * @param headers The headers which will be added to the request. For example: new HashMap<String, Object>() {{ put("Content-Type", "application/octet-stream"); }} Can be null.
+     * @param data The data which will put into the HTTP body of the request. Can be null.
+     * @param callback A {@link com.cloudmine.api.rest.callbacks.CMSocialLoginResponseCallback which has a {@link com.cloudmine.api.rest.response.CMSocialLoginResponse}.
+     * @throws InvalidRequestException A {@link com.cloudmine.api.exceptions.InvalidRequestException} is thrown if you do not use an appropriate HTTP verb. Note that PATCH is not support as of now.
+     */
+    public void asyncSocialGraphQueryOnNetwork(CMSocial.Service service,
+                                               HttpVerb httpVerb,
+                                               String baseQuery,
+                                               HashMap<String, Object> parameters,
```
Comment (EmilyMcCarthy) [inline src/main/java/com/cloudmine/api/rest/UserCMWebService.java]: Map

File: src/main/java/com/cloudmine/api/rest/UserCMWebService.java [hunk 7]
```diff
@@ -111,6 +111,66 @@ public void asyncLoadAccessLists(Callback<CMObjectResponse> callback) {
         executeAsyncCommand(get, callback, cmObjectResponseConstructor());
     }
 
+    /**
+     * A call to execute the Query on the Given Social Network. Returns the response body in the Callback.
```
Comment (EmilyMcCarthy) [inline src/main/java/com/cloudmine/api/rest/UserCMWebService.java]: Given is not a proper noun; really, neither is Social Network

File: src/main/java/com/cloudmine/api/rest/UserCMWebService.java [hunk 8]
```diff
@@ -111,6 +111,66 @@ public void asyncLoadAccessLists(Callback<CMObjectResponse> callback) {
...
+     *
+     * @param service The CMSocial.Service service you
```
Comment (EmilyMcCarthy) [inline src/main/java/com/cloudmine/api/rest/UserCMWebService.java]: the service I what? Don't leave me hanging :(

File: src/main/java/com/cloudmine/api/rest/UserCMWebService.java [hunk 9]
```diff
@@ -111,6 +111,66 @@ public void asyncLoadAccessLists(Callback<CMObjectResponse> callback) {
...
+     * @param httpVerb The {@link com.cloudmine.api.rest.HttpVerb} for the request.
+     * @param baseQuery The base query for the URL. Can be an empty string.
```
Comment (EmilyMcCarthy) [inline src/main/java/com/cloudmine/api/rest/UserCMWebService.java]: If this can be empty, also allow it to be null

File: src/main/java/com/cloudmine/api/rest/UserCMWebService.java [hunk 10]
```diff
@@ -111,6 +111,66 @@ public void asyncLoadAccessLists(Callback<CMObjectResponse> callback) {
...
+     * @param parameters The Parameters which will be added to the request. For example: new HashMap<String, Object>() {{ put("Param1", "Value"); }} Can be null.
+     * @param headers The headers which will be added to the request. For example: new HashMap<String, Object>() {{ put("Content-Type", "application/octet-stream"); }} Can be null.
```
Comment (EmilyMcCarthy) [inline src/main/java/com/cloudmine/api/rest/UserCMWebService.java]: You don't need to explain how to add values to a Map; also some people will be weird out by your double block initialization. Instead just say it is a mapping from header name to header value

File: src/main/java/com/cloudmine/api/rest/UserCMWebService.java [hunk 11]
```diff
@@ -111,6 +111,66 @@ public void asyncLoadAccessLists(Callback<CMObjectResponse> callback) {
...
+     * @param data The data which will put into the HTTP body of the request. Can be null.
+     * @param callback A {@link com.cloudmine.api.rest.callbacks.CMSocialLoginResponseCallback which has a {@link com.cloudmine.api.rest.response.CMSocialLoginResponse}.
+     * @throws InvalidRequestException A {@link com.cloudmine.api.exceptions.InvalidRequestException} is thrown if you do not use an appropriate HTTP verb. Note that PATCH is not support as of now.
```
Comment (EmilyMcCarthy) [inline src/main/java/com/cloudmine/api/rest/UserCMWebService.java]: "PATCH is not support as of now" is not proper english

Comment (EmilyMcCarthy) [inline src/main/java/com/cloudmine/api/rest/UserCMWebService.java]: Also, the note about PATCH being invalid should probably be on the httpVerb line

File: src/main/java/com/cloudmine/api/rest/response/SocialGraphResponse.java [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+        super(response);
+    }
+
+    /**
+     * Do not use. Use getStatusCode() instead.
+     * @return null.
+     */
+    @Override
+    public SocialGraphCode getResponseCode() {
+        return null;
```
Comment (EmilyMcCarthy) [inline src/main/java/com/cloudmine/api/rest/response/SocialGraphResponse.java]: Don't return null here; return a social status code that indicates it is not a valid status code, otherwise you break the methods contract

File: src/main/java/com/cloudmine/api/rest/response/code/SocialGraphCode.java [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+    /**
+     * Get the HTTP status code this enum represents
+     * @return
+     */
+    public int getStatusCode() {
+        return statusCode;
+    }
+
+    public static SocialGraphCode codeForStatus(int statusCode) {
+        return null;
```
Comment (EmilyMcCarthy) [inline src/main/java/com/cloudmine/api/rest/response/code/SocialGraphCode.java]: Don't return null here, and definitely never return null without indicating it is a possibility in your code.

File: src/main/java/com/cloudmine/api/rest/response/code/SocialGraphCode.java [hunk 14]
```diff
@@ -0,0 +1,27 @@
+package com.cloudmine.api.rest.response.code;
+
+/**
+ * <br>
+ * Copyright CloudMine LLC. All rights reserved<br>
+ * See LICENSE file included with SDK for details.
+ */
+public enum SocialGraphCode {
+    OK(200);
```
Comment (EmilyMcCarthy) [inline src/main/java/com/cloudmine/api/rest/response/code/SocialGraphCode.java]: Add a code for FAILURE for anything not in the 200s

File: src/test/java/com/cloudmine/api/integration/SocialIntegrationTest.java [hunk 15]
```diff
@@ -0,0 +1,288 @@
...
+ */
+public class SocialIntegrationTest extends ServiceTestBase {
+
+    private static final CMUser socialUser = new CMUser("ethan@cloudmine.me", "testing");
+    private static UserCMWebService socialService;
+    private String gistIdentifier;
+    private String gistTestingId;
+
+    @Before
+    public void setUp() {
```
Comment (EmilyMcCarthy) [inline src/test/java/com/cloudmine/api/integration/SocialIntegrationTest.java]: Is there a reason the ServiceTestBase setUp cannot be used here?

File: src/test/java/com/cloudmine/api/integration/SocialIntegrationTest.java [hunk 16]
```diff
@@ -0,0 +1,288 @@
...
+        reset();
+        CMApiCredentials.initialize("f2bc46fbb67948cea6dbed4f6f39940f", "c826d8bcad8a4f8685cef78efeb21ec3");
+
+    }
+
+    @Test
+    public void testFIRST() {
```
Comment (EmilyMcCarthy) [inline src/test/java/com/cloudmine/api/integration/SocialIntegrationTest.java:43]: Testing that a user can login is tested in other places in the code, why is this necessary?

File: src/test/java/com/cloudmine/api/integration/SocialIntegrationTest.java [hunk 17]
```diff
@@ -0,0 +1,288 @@
...
+            }
+        }));
+
+        waitForTestResults();
+        Assert.assertNotNull(socialUser.getSessionToken());
+    }
+
+    @Test
+    public void testTwitter() {
+        reset();
```
Comment (EmilyMcCarthy) [inline src/test/java/com/cloudmine/api/integration/SocialIntegrationTest.java]: Only need to call reset if you have already called waitForTestResults inside this method already

File: src/test/java/com/cloudmine/api/integration/SocialIntegrationTest.java [hunk 18]
```diff
@@ -0,0 +1,288 @@
...
+        socialService.asyncSocialGraphQueryOnNetwork(
+                CMSocial.Service.TWITTER,
+                HttpVerb.GET,
+                "statuses/home_timeline.json",
+                new HashMap<String, Object>() {{ put("include_entities", Boolean.valueOf(true)); }},
+                null,
```
Comment (EmilyMcCarthy) [inline src/test/java/com/cloudmine/api/integration/SocialIntegrationTest.java:74]: If there is a method that will often be called with nulls passed in for some of the arguments, just create a different method with the same name that is missing those arguments

File: src/test/java/com/cloudmine/api/integration/SocialIntegrationTest.java [hunk 19]
```diff
@@ -0,0 +1,288 @@
...
+                null,
+                testCallback(new SocialGraphCallback() {
+                    public void onCompletion(SocialGraphResponse response) {
+                        assertTrue(response.wasSuccess());
+                    }
+
+                    public void onFailure(Throwable t, String msg) {
+                        Assert.fail("Failed To Get Timeline!");
```
Comment (EmilyMcCarthy) [inline src/test/java/com/cloudmine/api/integration/SocialIntegrationTest.java]: If onFailure is called inside a testCallback the test is considered failed already

File: src/test/java/com/cloudmine/api/integration/SocialIntegrationTest.java [hunk 20]
```diff
@@ -0,0 +1,288 @@
...
+                    }
+                }
+                ));
+
+        waitForTestResults();
+    }
+
+    @Ignore //We don't need more tweets. this works.
```
Comment (EmilyMcCarthy) [inline src/test/java/com/cloudmine/api/integration/SocialIntegrationTest.java]: Can we create some fake accounts to run these tests on? Not much use having Integration tests that never run

File: src/test/java/com/cloudmine/api/integration/SocialIntegrationTest.java [hunk 21]
```diff
@@ -0,0 +1,288 @@
...
+                    }
+                }
+                ));
+        waitForTestResults();
+    }
+
+    @Test
+    public void testDropboxUpload() {
+        try {
+            byte[] bytes = readFileToByteArray(new File("/Users/ethan/dropbox_upload_test.txt"));
```
Comment (EmilyMcCarthy) [inline src/test/java/com/cloudmine/api/integration/SocialIntegrationTest.java]: I get the failing this test is going to fail when I run it on my machine

Comment (EmilyMcCarthy) [inline src/test/java/com/cloudmine/api/integration/SocialIntegrationTest.java]: freudian slip... feeling that it will fail

File: src/test/java/com/cloudmine/api/integration/SocialIntegrationTest.java [hunk 22]
```diff
@@ -0,0 +1,288 @@
...
+                            System.out.println("Code2: " + response.getStatusCode());
+                            assertTrue(response.wasSuccess());
+                        }
+
+                        public void onFailure(Throwable t, String msg) {
+                            Assert.fail("Failed To Upload Dropbox File");
+                        }
+                    }
+                    ));
+        } catch (Exception e) {
```
Comment (EmilyMcCarthy) [inline src/test/java/com/cloudmine/api/integration/SocialIntegrationTest.java:252]: You don't need to do this; if an uncaught exception occurs during a test case, the test case fails

File: src/test/java/com/cloudmine/api/integration/SocialIntegrationTest.java [hunk 23]
```diff
@@ -0,0 +1,288 @@
...
+            socialService.asyncSocialGraphQueryOnNetwork(
+                    CMSocial.Service.DROPBOX,
+                    HttpVerb.PUT,
+                    "files_put/dropbox/dropbox_upload_test.txt",
+                    null,
+                    new HashMap<String, Object>() {{ put("Content-Type", "application/octet-stream"); put("Content-Length", "" + length); }},
+                    file,
+                    testCallback(new SocialGraphCallback() {
+                        public void onCompletion(SocialGraphResponse response) {
+                            System.out.println("Code2: " + response.getStatusCode());
```
Comment (EmilyMcCarthy) [inline src/test/java/com/cloudmine/api/integration/SocialIntegrationTest.java]: use the logger for debug statements, but this can probably juse be removed entirely

File: src/main/java/com/cloudmine/api/rest/response/SocialGraphResponse.java [hunk 24]
```diff
@@ -0,0 +1,44 @@
...
+import java.io.StringWriter;
+
+/**
+ * Holds the response to a social graph query.
+ * <strong>Important:</strong> Use response.getStatusCode() for the HTTP Code, not response.getResponseCode(), because the code could be ANY int value.
+ * <br>
+ * Copyright CloudMine LLC. All rights reserved<br>
+ * See LICENSE file included with SDK for details.
+ */
+public class SocialGraphResponse extends ResponseBase<SocialGraphCode> {
```
Comment (EmilyMcCarthy) [inline src/main/java/com/cloudmine/api/rest/response/SocialGraphResponse.java:21]: Add a method to return the response as a SimpleCMObject

File: src/main/java/com/cloudmine/api/rest/response/SocialGraphResponse.java (hunk 12 above)
Comment (ethanmick) [inline src/main/java/com/cloudmine/api/rest/response/SocialGraphResponse.java]: Check this out in the new commit, I'm not sure if it's what you meant.

File: src/test/java/com/cloudmine/api/integration/SocialIntegrationTest.java (hunk 16 above)
Comment (ethanmick) [inline src/test/java/com/cloudmine/api/integration/SocialIntegrationTest.java:43]: It's not really a test - it's a setup so I can use the user in the subsequent tests. But the @Before directive runs the block before each test, and the user doesn't not need to be logged in each time - just once.

File: src/test/java/com/cloudmine/api/integration/SocialIntegrationTest.java (hunk 20 above)
Comment (ethanmick) [inline src/test/java/com/cloudmine/api/integration/SocialIntegrationTest.java]: Yess..... but that's a lot of work.

File: src/main/java/com/cloudmine/api/rest/response/SocialGraphResponse.java (hunk 24 above)
Comment (ethanmick) [inline src/main/java/com/cloudmine/api/rest/response/SocialGraphResponse.java:21]: Check to see if I did this correctly.

File: src/main/java/com/cloudmine/api/rest/response/code/SocialGraphCode.java (hunk 14 above)
Comment (ethanmick) [inline src/main/java/com/cloudmine/api/rest/response/code/SocialGraphCode.java]: Check to see how I did this.

File: src/main/java/com/cloudmine/api/rest/response/code/SocialGraphCode.java [hunk 25]
```diff
@@ -0,0 +1,40 @@
...
+     * Get the HTTP status code this enum represents
+     * @return
+     */
+    public int getStatusCode() {
+        return statusCode;
+    }
+
+    public static SocialGraphCode codeForStatus(int statusCode) {
+        for(SocialGraphCode code : SocialGraphCode.values()) {
+            if(statusCode <= code.upperRange && statusCode >= code.lowerRange ) {
```
Comment (EmilyMcCarthy) [inline src/main/java/com/cloudmine/api/rest/response/code/SocialGraphCode.java]: do: code.lowerRange <= statusCode && statusCode < upperRange
For one it is easier to read, for another a 300 code should not return QUERY_SUCCESS; as written what gets returned for status 300 depends on the iteration order

File: src/main/java/com/cloudmine/api/rest/response/SocialGraphResponse.java [hunk 26]
```diff
@@ -0,0 +1,49 @@
...
+    };
+
+
+    public SocialGraphResponse(HttpResponse response) {
+        super(response);
+    }
+
+    /**
+     * Do not use. Use getStatusCode() instead.
+     * @return null.
```
Comment (EmilyMcCarthy) [inline src/main/java/com/cloudmine/api/rest/response/SocialGraphResponse.java]: no longer returns null

File: src/main/java/com/cloudmine/api/rest/response/SocialGraphResponse.java [hunk 27]
```diff
@@ -0,0 +1,49 @@
...
+     */
+    @Override
+    public SocialGraphCode getResponseCode() {
+        return SocialGraphCode.codeForStatus(getStatusCode());
+    }
+
+    public SimpleCMObject toSimpleCMObject() {
+        return new SimpleCMObject(this);
```
Comment (EmilyMcCarthy) [inline src/main/java/com/cloudmine/api/rest/response/SocialGraphResponse.java]: This works, but I would use return new SimpleCMObject(new TransportableString(getMessageBody()));
This will skip converting the Map that represents this response to a JSON string which is just going to be converted back to a SimpleCMObject

Comment (ethanmick) [inline src/main/java/com/cloudmine/api/rest/response/SocialGraphResponse.java]: What happens if getMessageBody() returns XML, instead of JSON? Our SimpleCMObject will fail being created?
